text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+OTLP data with a very slight advantage for Zlib. The same compression algorithms have been tested on our new
+"column-oriented" representation. Zlib and Zstd are the two best compression algorithms for OTLP Arrow.
+
+> As this plot shows, the columnar representation is almost twice as efficient in terms of bandwidth gain. However, it
+> is
+> possible to do much better by taking advantage of the native support for multivariate time-series in OTLP Arrow.
+
+The following graph compares the best OTLP compression algorithm (Zlib) with OTLP Arrow + Zstd + multivariate time-series
+support.
+
+![Multivariate metrics bytes](img/0156_multivariate_metrics_bytes.png)
+
+> With this optimization, OTLP Arrow is a little over 4 times more efficient than OTLP. The zstd compression algorithm
+> stands out for columnar data.
+
+Even for small batches of metrics, the OTLP Arrow representation is more efficient than the OTLP representation in a
+*multivariate time-series context* (see table below).
+
+![Metrics small batches](img/0156_metrics_small_batches.png)
+
+#### Serialization/Compression Time Analysis
+
+This section compares the time spent creating, serializing, compressing, decompressing and deserializing batches of
+metrics with OTLP and OTLP Arrow while looking to optimize message size (Zlib for OTLP and Zstd for OTLP Arrow). In both
+cases, the data producer delivers the metrics in a row-oriented approach. The Arrow OTLP implementation therefore
+integrate the
+row-to-column conversion time.
+
+The following stacked bar plot compares side by side the distribution of time spent for each step and for each version
+of the protocol.
+
+![Breakdown of time per step for each batch size and each protocol](img/0156_metrics_step_times.png)
+
+When cumulating the times of all steps, OTLP Arrow (with multivariate time-series encoding) is 3 times faster than
+standard OTLP.
+
+The batch creation phase for OTLP Arrow represents almost all the time spent. Serialization and deserialization times
+are
+almost zero due to the use of Flatbuffer by Apache Arrow (ser/deser without parsing/unpacking). Compression and
+decompression
+times are extremely low due to the fact that the size of the Arrow OTLP message before compression is more than 30 times
+smaller than an OTLP message with the same content (see previous table).
+
+> It should be possible to significantly optimize the creation of OTLP Arrow batches for contexts where the structure
+> of the metrics to be sent is known in advance.
+
+### Logs
+
+#### Batch Size Analysis
+
+The dataset used for this benchmark contains 200K logs with 43 different structures (i.e. combination of different
+fields).
+
+The following multi-line plot show the compressed batch size in bytes for different protocols and combination of
+parameters:
+
+* protocol (OTLP or OTLP Arrow)
+* batch size (5000, 10000, 25000, 50000, 100000)
```

Comment:
It would be very interesting to see smaller batch sizes too, e.g. 1, 10, 100.

Reply:
I'm working on it. Some of these results are already visible in the README.md page of the reference implementation. Right now, without batching OTLP Arrow is not efficient at all because of some constant overhead during the batch creation step. This can probably be improved significantly by reusing some constructs more (e.g. arrow.RecordBuilder) but we will probably never be faster than a row-oriented approach when there is no batching.